src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -55,6 +58,11 @@ SendCommandToWorkersAsUser(TargetWorkerSet targetWorkerSet, const char *nodeUser
 	List *workerNodeList = TargetWorkerSetNodeList(targetWorkerSet, ShareLock);
 	ListCell *workerNodeCell = NULL;
 
+	if (targetWorkerSet == WORKERS_WITH_METADATA)
+	{
+		ErrorIfAnyMetadataNodeOutOfSync(workerNodeList);
```

Comment:
`SendCommandToWorkersAsUser` doesn't seem like quite the right place to put this logic. I wonder whether we should be wrapping these functions into something like "SendCommandToWorkersWithMetadata" to have a clear separation of concerns.

Reply:
I'm considering another alternative. Put the check to `SendRemoteCommandParams() / SendRemoteCommand()`.

Moving a similar check to those functions requires reading the worker node from workernodeash on every command, which might be expensive to do. 

So, I'm considering to keep a global state of `bool HasWorkerWithOutOfSyncMetadata`, and in `SendRemoteCommandParams() / SendRemoteCommand()` we can do:
```C
if (HasWorkerWithOutOfSyncMetadata)
{
    WorkerNode *w = ...; /* only read the worker when necessary */
     if (!w-> metadataSynced)
        error
}
```

In this patch, keeping track of where to add `ErrorIfAnyMetadataNodeOutOfSync() ` is hard, in my proposal maintaining `HasWorkerWithOutOfSyncMetadata` is slightly hard, but at least we'd be ensuring any future command sending API is safe to use.

What do you think for that?

Reply:
My problem is mainly that these are generic functions for sending commands and we're putting logic related to metadata syncing into them. I'm proposing we do some refactoring to instead call wrappers that are specific to metadata syncing.